### PR TITLE
NO-TICKET: merge master to dev

### PR DESCRIPTION
**Pull request by Fraser999**
This PR merges current `master` into current `dev`.

The [fourth last commit](https://github.com/CasperLabs/casper-node/commit/48ca022fd0085d09f6b61134c3b72c9de7076929) is the actual merge and is probably the only one worth reviewing.

### CODE REVIEW

File: node/src/components/linear_chain/event.rs [hunk 1]
```diff
@@ -44,6 +46,9 @@ impl<I: Display> Display for Event<I> {
             Event::NewLinearChainBlock { block, .. } => {
                 write!(f, "linear chain new block: {}", block.hash())
             }
+            Event::KnownLinearChainBlock(block) => {
```
Comment (afck) [inline node/src/components/linear_chain/event.rs:49]: It looks like that variant isn't constructed anywhere anymore. Should it be removed?

File: node/src/components/storage/tests.rs [hunk 2]
```diff
@@ -716,7 +716,6 @@ fn store_random_execution_results() {
 }
 
 #[test]
-#[should_panic(expected = "duplicate execution result")]
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply node/src/components/storage/tests.rs:719]: It's arguably of limited benefit, but it does at least now show that overwriting execution results with different results doesn't cause a panic.

File: node/src/components/storage.rs [hunk 3]
```diff
@@ -528,17 +519,13 @@ impl Storage {
...
                     if let Some(prev) = metadata.execution_results.get(&block_hash) {
-                        if prev != &execution_result {
-                            return Err(Error::DuplicateExecutionResult {
-                                deploy_hash,
-                                block_hash: *block_hash,
-                            });
+                        if prev == &execution_result {
+                            continue;
+                        } else {
+                            debug!(%deploy_hash, %block_hash, "different execution result");
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply node/src/components/storage.rs:527]: This was covered in https://github.com/CasperLabs/casper-node/pull/1167.  We believe it is safe.  In the case which tripped us up, the results were the same modulo ordering.

If we want to change from this behaviour, I think it should be ticketed outside of this PR.

**Review by afck — APPROVED** [review]

**Review by sacherjj — APPROVED** [review]

**Comment by sacherjj** [comment]
bors merge

**Pull request merged**